service-base/src/main/java/org/eclipse/hono/service/registration/AbstractRegistrationService.java

```diff
@@ -210,40 +235,63 @@ public void assertRegistration(
      */
     protected boolean isGatewayAuthorized(final String gatewayId, final JsonObject gatewayData,
             final String deviceId, final JsonObject deviceData) {
-
         Objects.requireNonNull(gatewayId);
         Objects.requireNonNull(gatewayData);
         Objects.requireNonNull(deviceId);
         Objects.requireNonNull(deviceData);
 
-        final Object obj = deviceData.getValue(RegistrationConstants.FIELD_VIA);
-        if (obj instanceof JsonArray) {
-            // get the first matching entry
-            return ((JsonArray) obj)
-                    .stream()
-                    .filter(String.class::isInstance)
-                    .anyMatch(gatewayId::equals);
-        } else if (obj instanceof String) {
-            // compare the string directly
-            return gatewayId.equals(obj);
+        final JsonArray via = convertObjectToJsonArray(deviceData.getValue(RegistrationConstants.FIELD_VIA));
+        final JsonArray viaGroups = convertObjectToJsonArray(deviceData.getValue(RegistrationConstants.FIELD_VIA_GROUPS));
+        final JsonArray memberOf = convertObjectToJsonArray(gatewayData.getValue(RegistrationConstants.FIELD_MEMBER_OF));
+
+        return isGatewayInVia(gatewayId, via) || isGatewayInViaGroups(memberOf, viaGroups);
+    }
+
+    private boolean isGatewayInVia(final String gatewayId, final JsonArray via) {
+        return via
+                .stream()
+                .filter(String.class::isInstance)
+                .anyMatch(gatewayId::equals);
+    }
+
+    private boolean isGatewayInViaGroups(final JsonArray memberOf, final JsonArray viaGroups) {
```

Comment:
Some renaming would make this more understandable, IMHO. (At first sight, I thought the gatewayId is searched for in the viaGroups here). E.g.
isGatewayInViaGroups -> isGatewayGroupInViaGroups
memberOf -> gatewayDataMemberOf or gatewayGroups or gatewayDataGroups (also in calling method)